RetailCoder.VBE/Navigation/CodeMetrics/CodeMetricsAnalyst.cs

```diff
@@ -0,0 +1,194 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Rubberduck.Parsing.VBA;
+using Antlr4.Runtime.Tree;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.VBEditor;
+using Antlr4.Runtime.Misc;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.SmartIndenter;
+
+namespace Rubberduck.Navigation.CodeMetrics
+{
+    public class CodeMetricsAnalyst : ICodeMetricsAnalyst
+    {
+        private readonly IIndenterSettings indenterSettings;
+    
+        public CodeMetricsAnalyst(IIndenterSettings indenterSettings)
+        {
+            this.indenterSettings = indenterSettings;
+        }
+
+        public IEnumerable<ModuleMetricsResult> ModuleMetrics(RubberduckParserState state)
+        {
+            if (state == null || !state.AllUserDeclarations.Any())
+            {
+                // must not return Enumerable.Empty
+                yield break;
+            }
+
+            var trees = state.ParseTrees;
+            var results = new List<CodeMetricsResult>();
+
+            foreach (var moduleTree in trees)
+            {
+                yield return GetModuleResult(moduleTree.Key, moduleTree.Value, state.DeclarationFinder);
+            };
+        }
+
+        public ModuleMetricsResult GetModuleResult(RubberduckParserState state, QualifiedModuleName qmn)
+        {
+            return GetModuleResult(qmn, state.GetParseTree(qmn), state.DeclarationFinder);
+        }
+
+        private ModuleMetricsResult GetModuleResult(QualifiedModuleName qmn, IParseTree moduleTree, DeclarationFinder declarationFinder)
+        {
+            // Consider rewrite as visitor? That should make subtrees easier and allow us to expand metrics
```

Comment:
Note that a visitor can be cancelled whether a walker can't be as easily cancelled. That may have some UI ramifications if there can be a delay when interacting with the UI. If it executes quickly and UI is responsive, then it might be just be fine as a walker.

Reply:
sure can be. Just check a CancellationToken when entering any rule. Bit of a pain though, so this usually isn't done...